docs/rfcs/0000-persistent-logging-component.md

```diff
@@ -0,0 +1,111 @@
+---
+title: Persistent logging component
+layout: page
+permalink: /rfc/0000-persistent-logger
+---
+
+* Start date: 2020-10-10
+* RFC PR: [#0000](https://github.com/mozilla-mobile/android-components/pull/0000)
+
+## Summary
+Introduce a component for recording application logs on a device to a persistent store that can be shared with developers for debugging.
+
+## Motivation
+Reproducing bugs that a user encounters in the wild is a difficult process for the developer to get their device in a similar state in order to debug. We have useful information in our Android logs today that aid in debugging that we also do not have access to, but give us insight into the app state.
+
+With a component that logs to disk for later, we can follow changes of state within the browser and crash info that can be shared with developers or SUMO to allow for more useful debugging.
+
+Asking users for `adb logcat` output is a common step when trying to debug certain kinds of problems (such as account, push, sync, etc). However, for non-technical users this could be tricky, and for technical users this request assumes presence of `adb`, or willingness to install it. The easier we can make log extraction, the more accessible this process becomes for a diverse audience, helping us build a more robust application.
+
+## Guide-level explanation
+
+One requirement that we wanted to introduce is that logging should not happen in the background unless explicitly turned on. Some advantages for this:
+ - We want to target reproducible bugs, and not random crashes, which we have a component for that (`lib-crash`).
+ - There is less IO done on production devices that are running just fine.
+   - Performing on-demand recording allows us to vastly simplify how we process the logs.
+   - The UI becomes more intuitive, since we now have a recording unit which can be viewed, shared, deleted, etc. A simple stream of log statements doesn't have similar affordances.
+ - We can avoid many maintainence tasks like an ever growing database of logs that need to be flushed.
+ - The user can be explicit with the scope of what is collected.
+
+### Recording
+
+> **Note**: We use the term "recording" here to mean an event which allows logs to be stored to the database that begin with a 'start' action and ends with a 'stop' action from the user or the app crashes. Each 'start' generates a new recording ID. This is a similar concept to a screen recording of a bug being reproduced in a GIF.
+
+In order to do this, we want to introduce a persistent notification service that would be show up only when enabled, similar to how USB debugging requires explicit action to enable. When a bug is ready to be reproduced, the user can click on a 'Record' button in the notification, perform the steps to reproduce the bug, and click the 'Stop' button right after. To capture, a 'Record after on startup' option will be possible as well to capture startup logging.
```

Comment:
Yeah, this would be easier. I imagine that we would be asking more users via SUMO as well to enable logging for us so that we can debug issues, so those users may not always notice/remember to turn this off. We could also reduce our chances of maxing out the database if someone (me) were to just forget to turn it off and/or having to consider auto-cleanup strategies.

Reply:
@grigoryk what you think?

Reply:
I think that's ok, it's mostly the same thing except with a simpler UI. We retain an explicit action from the user as a necessary trigger, notification UI is simpler, and we get an easy way to do startup log recording.

We could still retain the concept of a recording at the schema level, to make it easier to manage these recordings (share, clean up old ones, etc). It will map well to the user actions (start, stop).